File: C/insertion_sort.c
#include <stdio.h>

/*Sort the array using the insertion sort algorithm*/
void insertionSort(int arr[], int size)
{
    for(int i = 1; i < size; i++)
    {
        int key = arr[i];
        int j = i - 1;

        while (j >= 0 && arr[j] > key)
        {   
            arr[j + 1] = arr[j];
            j--;
        }
        arr[j + 1] = key;
    }
}

/*Print out the array into the console*/
void printArray(int arr[], int size) 
{
    for (int i = 0; i < size; i++) 
    {
        printf("%d ", arr[i]);
    }
    printf("\n");
}

/*Sorts and prints out the given array before and after sorting*/
void sortArray(int arr[], int size)
{
    // Print before sorting
    printf("Before sorting: ");
    printArray(arr, size);

    insertionSort(arr, size); // Sort array
    
    // Print after sorting
    printf("After sorting: ");
    printArray(arr, size);
    printf("\n");
}

int main()
{
    //First array
    int arr[] = {34, 56, 11, 21, 9};
    int size = sizeof(arr) / sizeof(arr[0]);
    sortArray(arr, size);
    
    //Second Array
    int arr2[] = {7, 23, 45, 12, 38, 6, 29, 50, 18, 33};
    int size2 = sizeof(arr2) / sizeof(arr2[0]);
    sortArray(arr2, size2);

    //Third Array
    int arr3[] = {5, 12, 47, 29, 33, 8, 20, 44, 17, 3, 
                50, 25, 6, 39, 14, 41, 9, 22, 36, 48};
    int size3 = sizeof(arr3) / sizeof(arr3[0]);
    sortArray(arr3, size3);
    
    return 0;
}

File: Java/insertion_sort.java
/* Function to sort array using insertion sort */
public class insertion_sort {
    void sort(int arr[])
    {
        int n = arr.length;
        for (int i = 1; i < n; ++i) {
            int key = arr[i];
            int j = i - 1;

            while (j >= 0 && arr[j] > key) {
                arr[j + 1] = arr[j];
                j = j - 1;
            }
            arr[j + 1] = key;
        }
    }

    /* A utility function to print array of size n */
    static void printArray(int arr[])
    {
        int n = arr.length;
        for (int i = 0; i < n; ++i)
            System.out.print(arr[i] + " ");

        System.out.println();
    }

    // Driver method
    public static void main(String args[])
    {
        int arr[] = { 12, 11, 13, 5, 6 };
	int arr2[] = { 10, 8, 6, 4, 2, 7 };
	int arr3[] = { 4, 12, 0, 21, 105, 17 };

        insertion_sort ob = new insertion_sort();

	System.out.print("Test case 1: ");
	printArray(arr);
	System.out.print("Sorted: ");
	ob.sort(arr);
	printArray(arr);
	System.out.println();

	System.out.print("Test case 2: ");
	printArray(arr2);
	System.out.print("Sorted: ");
	ob.sort(arr2);
	printArray(arr2);
	System.out.println();

	System.out.print("Test case 3: ");
	printArray(arr3);
	System.out.print("Sorted: ");
	ob.sort(arr3);
        printArray(arr3);
    }
}
